torch/csrc/jit/graph_executor.cpp

```diff
@@ -305,137 +329,153 @@ struct GraphExecutorImpl {
-        return true;
-    }
-    return false;
-  }
-
-  const Code & getOrCreateAutogradFallback() {
+  const ExecutionPlan & getOrCompileFallback() {
     std::lock_guard<std::mutex> lock(compile_mutex);
-    if(autograd_fallback) {
-      return autograd_fallback;
-    }
-    auto graph_ = graph->copy();
-    runRequiredPasses(graph_);
-    if(optimize) {
-      runOptimization(graph_, /*graphMustSupportVariables=*/true);
-      if(!isDifferentiable(*graph_)) {
-        EraseShapeInformation(*graph_);
-        CreateAutodiffSubgraphs(*graph_);
-      }
+    if(!fallback) {
+      auto graph_ = graph->copy();
+      runRequiredPasses(graph_);
+      fallback = ExecutionPlan(graph_);
     }
-    autograd_fallback_graph = graph_;
-    autograd_fallback = Code(graph_);
-    return autograd_fallback;
+    return fallback;
   }
-  const ExecutionPlan & getOrCompile(at::ArrayRef<IValue> inputs) {
+
+  const ExecutionPlan & getOrCompile(const Stack& stack) {
     // outside lock guard, to minimize the time holding the lock on the fast path
     // ArgumentSpec even computes its hashCode here.
-    ArgumentSpec spec(autograd::GradMode::is_enabled(), inputs);
+    ArgumentSpec spec(autograd::GradMode::is_enabled(), last(stack, num_inputs));
     {
       std::lock_guard<std::mutex> lock(compile_mutex);
       auto it = plan_cache.find(spec);
-      if(it != plan_cache.end())
+      if (it != plan_cache.end())
         return it->second;
       auto plan = compileSpec(spec);
       auto r = plan_cache.emplace(std::move(spec), std::move(plan));
       return r.first->second;
     }
   }
 
-  bool argumentSpecRequiresGradient(const ArgumentSpec & spec) {
+  ExecutionPlan compileSpec(const ArgumentSpec & spec) {
+    auto opt_graph = graph->copy();
+
+    // Phase 1. Specialize to input definedness (this is very important for
+    //          gradient graphs), and run required passes to bring the graph
+    //          to an executable form.
+    specializeGrad(opt_graph, spec);
+    runRequiredPasses(opt_graph);
+
+    // Phase 2. Propagate detailed information about the spec through the
+    //          graph (enabled more specializations in later passes).
+    PropagateInputShapes(*opt_graph, spec);
+
+    // Phase 3. Run differentiable optimizations (i.e. simple graph rewrites that
+    //          we can still execute using autograd).
+    runOptimization(opt_graph, spec);
+
+    // Phase 4. If this graph will be differentiated, we need to slice out the
+    //          symbolically differentiable subgraphs for further optimizations.
+    // Phase 5. Apply non-differentiable optimizations to the graphs we've found
+    //          (or the whole grpah if we know we won't need its derivative).
+    if (needsGradient(opt_graph, spec)) {
+      auto diff_nodes = CreateAutodiffSubgraphs(*opt_graph);
+      for (Node * dnode : diff_nodes) {
+        // XXX: we don't have requires_grad information on the intermediate values,
```

Comment:
Just to double-check: we can fix this mostly with requires_grad propagation?

Reply:
Yes. It will require us to record this information on tensor types, but that will need to come in a follow up PR.